src/commonMain/kotlin/org/treeWare/model/operator/Difference.kt

```diff
@@ -0,0 +1,407 @@
+        if(stepsFromAdd <= 0) { //Case: In or above an added node.
+            val addParent = addStack.first()
+            connectValue(addParent, addEntity)
+        }
+        else{ //Case: Not in or above an added node
+            stepsFromAdd --
+        }
+
+        /**===== Handle the delete tree =====**/
+        val deleteEntity = deleteStack.removeFirst() as MutableEntityModel
+        if(stepsFromDelete <= 0) { //Case: In or above a deleted node
+            val deleteParent = deleteStack.first()
+            connectValue(deleteParent, deleteEntity)
+        }
+        else{ //Case: Not in or above a deleted node
+            stepsFromDelete --
+        }
+
+        /**===== Handle the update tree =====**/
+        val updateEntity = updateStack.removeFirst() as MutableEntityModel
+        if(stepsFromUpdate <= 0) { //Case: In or above a deleted node
+            val updateParent = updateStack.first()
+            connectValue(updateParent, updateEntity)
+        }
+        else{ //Case: Not in or above a deleted node
+            stepsFromUpdate --
+        }
+    }
+
+    override fun visitSingleField(leaderFieldList: List<SingleFieldModel?>): TraversalAction =
+        visitField(leaderFieldList)
+
+    override fun leaveSingleField(leaderFieldList: List<SingleFieldModel?>) {
+        leaveField(leaderFieldList)
+    }
+
+    override fun visitListField(leaderFieldList: List<ListFieldModel?>): TraversalAction =
+        visitField(leaderFieldList)
+
+    override fun leaveListField(leaderFieldList: List<ListFieldModel?>) {
+        leaveField(leaderFieldList)
+    }
+
+    override fun visitSetField(leaderFieldList: List<SetFieldModel?>): TraversalAction =
+        visitField(leaderFieldList)
+
+    override fun leaveSetField(leaderFieldList: List<SetFieldModel?>) {
+        leaveField(leaderFieldList)
+    }
+
+    override fun visitPrimitive(leaderValueList: List<PrimitiveModel?>): TraversalAction { //relevant to UPDATE
+        val lastPrimitive = leaderValueList.lastNotNullOf { it } //may change later to have first and last, for update
+
+        /**=====Check if it's absent in the original=====**/
+        val addParent = addStack.first() as MutableFieldModel
+        if((leaderValueList.first() == null) or isKeyField(addParent)) { //if the original doesn't have the entity:
+            if(leaderValueList.first() == null){
+                stepsFromAdd = 0 //everything above this node should be included
+                //println("New Primitive found")
+            }
+
+            val addPrimitive = newChildValue(addParent) as MutablePrimitiveModel
+            addPrimitive.copyValueFrom(lastPrimitive)
+        }
+
+        /**=====Check if it's absent in the new version=====**/
+        val deleteParent = deleteStack.first() as MutableFieldModel
+        if((leaderValueList.last() == null) or isKeyField(deleteParent) ){//if the new version doesn't have the entity:
+            if(leaderValueList.last() == null){
```

Comment:
nit: save the value of first()/last() in variables so that they are not called more than once. Not sure how expensive these functions are, but just in case they are expensive. Saving them in variables (`val`) also let's you name them as `old` and `new` and that helps you avoid some of the comments at the end of the if-statements.